File: haskell/034/Main.hs
{-
    ProjectEuler 034
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

import Data.Array.Unboxed
import Data.Char

main :: IO ()
main = print . sum $ filter test [10..7 * fact 9]

-- 9! = 362880
fact :: Integral a => a -> a
fact n = fact' 1 n
    where
        fact' p 0 = p
        fact' p n = fact' (p * n) (n - 1)

facts :: UArray Int Int
facts = listArray (0, 9) $ map fact [0..9]

test :: Int -> Bool
test n = itr 0 n
    where
        itr s 0 = s == n
        itr s m = itr (s + facts ! (m `mod` 10)) (m `div` 10)

File: ocaml/034/main.ml
(*
    ProjectEuler 034
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let factorials = [|1; 1; 2; 6; 24; 120; 720; 5040; 40320; 362880|]

let max_n = 362880 * 7

let sum_of_factorials_of_digits n =
    let rec itr s = function
        | 0 -> s
        | r -> let f = Array.get factorials (r mod 10) in
            itr (s + f) (r / 10)
    in itr 0 n

let _ =
    let rec itr s n =
        if n >= max_n then s
        else let t = sum_of_factorials_of_digits n in
            if t = n then itr (s + n) (n + 1)
            else itr s (n + 1)
    in
    let s = itr 0 3 in
    print_int s;
    print_newline ()
